### PR TITLE
Move API to autograder-api & dynamic loading of autograder-core and autograder-extra

### LINKED ISSUE #564: Make the `autograder-core` update itself
### Description

This is going to be fun to implement. I do not want to bump a release for everything using the autograder, every time a hot fix is made.

To solve this, we implement
- checking for new releases on github,
- downloading the latest jar
- somehow replace the currently loaded classes with the updated ones?

### PR DESCRIPTION
**Pull request by Feuermagier**
This PR implements dynamic loading of autograder-core and autograder-extra (and dependencies). This fixes #564, similar to #566, but with different tradeoffs. The main benefit is that this implementation relies less on fragile class loader tricks. However, changing an already loaded implementation from within a running process is not supported.

The PR introduces two new modules:

* `autograder-api`: Implements the core API to the Autograder, without a dependency on any other autograder module (nor spoon). Notably missing is `UploadedFile`, which is not required for the simple API exposed to downstream tools such as the Grading Tool. The core classes that users interact with (Linter, Problem, TempLocation, CodePosition) are converted to interfaces, and the old implementations are moved to XYImpl. autograder-core and autograder-extra always use these concrete implementations instead of the interfaces, since the interfaces only expose the functionality required for downstream use. Downstream users now have two options: 
    * Depend on autograder-core (and optionally autograder-extra) and directly instantiate the respective Impls. The tests and autograder-cmd do this. There is no reflection or download of code involved in this approach.
    * Don't depend on autograder-core, but use the dynamic loading utilities in autograder-api. These utilities can download the autograder-full JAR from GitHub, load the classes into the running process, and discover the Impls using reflection.
* `autograder-full`: This module contains no code and is only used to build a JAR release that can be downloaded by the dynamic loader.

### CODE REVIEW

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -58,4 +60,6 @@ jobs:
         if: startsWith(github.ref, 'refs/tags/')
         with:
           body: 'Minor Release (mostly for the automatic download of the jar file)'
-          files: autograder-cmd/target/autograder-cmd.jar
\ No newline at end of file
+          files: |
+              autograder-cmd/target/autograder-cmd.jar
+              autograder-full/target/autograder-full.jar
```
Comment (Luro02) [inline .github/workflows/build.yml:65]: Why do we need two binaries? What is in `autograder-full` that is not in `autograder-cmd`?

File: autograder-api/src/main/java/de/firemage/autograder/api/CodePosition.java [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+public interface CodePosition {
+    Path path();
+    int startLine();
+
+    int endLine();
+
+    int startColumn();
+
+    int endColumn();
+    String readCompilationUnit();
```
Comment (Luro02) [inline autograder-api/src/main/java/de/firemage/autograder/api/CodePosition.java:14]: It is difficult to understand what this does without a javadoc. How about `readSourceFile`?

File: autograder-core/src/main/java/de/firemage/autograder/core/CodeLinter.java [hunk 3]
```diff
@@ -22,11 +23,11 @@ public interface CodeLinter<T extends Check> {
      * @return a list of problems found in the submission
      * @throws IOException if an I/O error occurs
      */
-    List<Problem> lint(
+    List<ProblemImpl> lint(
```
Comment (Luro02) [inline autograder-core/src/main/java/de/firemage/autograder/core/CodeLinter.java:26]: Why is this returning the `ProblemImpl` now?

File: autograder-core/src/main/java/de/firemage/autograder/core/file/TempLocationImpl.java [hunk 4]
```diff
@@ -1,31 +1,33 @@
...
-import java.io.Closeable;
 import java.io.File;
 import java.io.IOException;
-import java.io.Serializable;
 import java.nio.file.Files;
 import java.nio.file.Path;
 import java.util.List;
 import java.util.Random;
 
-public record TempLocation(File tempLocation) implements Serializable, Closeable {
```
Comment (Luro02) [inline autograder-core/src/main/java/de/firemage/autograder/core/file/TempLocationImpl.java:14]: Why did you change the record to a class?

File: autograder-core/src/test/java/de/firemage/autograder/core/CheckTest.java [hunk 5]
```diff
@@ -91,7 +94,7 @@ Stream<DynamicTest> createCheckTest() throws URISyntaxException, IOException {
             folders = paths.toList();
         }
 
-        TempLocation tempLocation = TempLocation.random();
+        TempLocation tempLocation = new TempLocationImpl();
```
Comment (Luro02) [inline autograder-core/src/test/java/de/firemage/autograder/core/CheckTest.java:97]: I liked the static method more

Comment (Feuermagier) [reply autograder-core/src/test/java/de/firemage/autograder/core/CheckTest.java:97]: The constructors are cleaner for reflection, so additional static methods felt redundant. But we can add them back in if you prefer.

File: .github/workflows/build.yml (hunk 1 above)
Comment (Feuermagier) [reply .github/workflows/build.yml:65]: Nothing. The intent was to strip unneeded dependencies and reduce the potential of clashing implementations at runtime, but I'm not sure if that's worth it

**Comment by Luro02** [comment]
I am quite unhappy that this PR renames the types to `ProblemImpl`/`TempLocationImpl` for the `autograder-api` interfaces. The interfaces are mostly for dynamic loading and are not used in the autograder-core/extra codebase.

My suggestion is to rename the interfaces in `autograder-api` and keep the original names for the `Problem`/`TempLocation` classes. That would make the number of changes a lot smaller and the code would look like before this PR.

Maybe `ProblemFacade`/`TempLocationFacade`?

File: autograder-core/src/test/java/de/firemage/autograder/core/CheckTest.java (hunk 5 above)
Comment (Luro02) [reply autograder-core/src/test/java/de/firemage/autograder/core/CheckTest.java:97]: I would like to have the factory method back.

File: .github/workflows/build.yml (hunk 1 above)
Comment (Luro02) [reply .github/workflows/build.yml:65]: I don't think that is worth it:

`autograder-cmd.jar`: 50.9MB
`autograder-full.jar`: 50.5MB

File: autograder-api/src/main/java/de/firemage/autograder/api/loader/AutograderLoader.java [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+                .classLoader(autograderClassLoader)
+                .instantiate();
+    }
+
+    public static TempLocation instantiateTempLocation() {
+        try {
+            Class.forName("de.firemage.autograder.core.file.TempLocationImpl", true, autograderClassLoader);
+        } catch (ClassNotFoundException e) {
+            throw new RuntimeException(e);
+        }
```
Comment (dfuchss) [inline autograder-api/src/main/java/de/firemage/autograder/api/loader/AutograderLoader.java:73]: What will that do? And why do you handle it different from something like Linter?

File: autograder-core/pom.xml [hunk 7]
```diff
@@ -11,14 +11,16 @@
...
+    <description>Core checks of the autograder</description>
     <url>https://github.com/Feuermagier/autograder/autograder-core</url>
 
-    <properties>
-        <fluent.version>0.70</fluent.version>
-    </properties>
-
     <dependencies>
+        <!-- Internal -->
+        <dependency>
```
Comment (dfuchss) [inline autograder-core/pom.xml:19]: Maybe you can set the dependency scope of the API to provided. As far as I know (but I have to check that) that would mean that the API is expected to be there and therefore it will not be directly included as transitive dependency.

-> You have to include it in autograder-cmd
-> You can omit it in autograder-full -> thus it will be excluded in the final JAR of autograder full -> only one version of the api will be loaded in intelligrade by construction (because the api is not included in the full jar)

File: autograder-extra/pom.xml [hunk 8]
```diff
@@ -21,6 +21,10 @@
 
     <dependencies>
         <!-- Internal -->
+        <dependency>
+            <groupId>de.firemage.autograder</groupId>
+            <artifactId>autograder-api</artifactId>
+        </dependency>
```
Comment (dfuchss) [inline autograder-extra/pom.xml:27]: Same here .. this would ensure that the API is not a transitive dependency.. You could also think about adding the extra jar to the provided artifacts and make a flag in the API to load it as well.

File: autograder-api/src/main/java/de/firemage/autograder/api/loader/AutograderLoader.java (hunk 6 above)
Comment (Feuermagier) [reply autograder-api/src/main/java/de/firemage/autograder/api/loader/AutograderLoader.java:73]: That was a leftover from debugging

**Comment by Feuermagier** [comment]
> I am quite unhappy that this PR renames the types to `ProblemImpl`/`TempLocationImpl` for the `autograder-api` interfaces. The interfaces are mostly for dynamic loading and are not used in the autograder-core/extra codebase.
> 
> My suggestion is to rename the interfaces in `autograder-api` and keep the original names for the `Problem`/`TempLocation` classes. That would make the number of changes a lot smaller and the code would look like before this PR.
> 
> Maybe `ProblemFacade`/`TempLocationFacade`?

I've changed it to `AbstractLinter`, `AbstractTempLocation`, `AbstractCodePosition`, and `AbstractProblem` (which is a bit dumb, because the 'concrete' `CodePosition` is itself abstract, but at least it's consistent).

File: .github/workflows/build.yml (hunk 1 above)
Comment (Feuermagier) [reply .github/workflows/build.yml:65]: Actually, the important reason is the thing that @dfuchss mentioned below: `autograder-full` may not contain `autograder-api`, or there will be clashes during dynamic loading. `autograder-cmd` on the other hand needs all dependencies.

**Comment by Luro02** [comment]
Why is `ProblemType` in api? I guess the reason is for type-safety?

I wouldn't expose the enum to downstream, only via Strings for the enum constants.
This should prevent problems with potential future changes and the code using
`autograder-api` does not instantiate the `ProblemType` variants explicitly? It should
only parse the config, maybe not even that. If type-safety is necessary, then expose a wrapper class like this:
```java
public record AutograderProblemType(String name) {}

// in `ProblemType` one could add a method to safely convert between them:
public enum ProblemType {
    ...;

    public AutograderProblemType toAutograderProblemType() {
        return new AutograderProblemType(this.toString());
    }
}
```

The current implementation would crash when new releases introduce new checks because these are not loaded via the class loader. It would be exceedingly difficult to replace the `ProblemType` enum with a different implementation, like making it an interface and having different implementations for `autograder-core` and `autograder-extra`.

The current implementation doesn't provide a good error-message when there are unknown `ProblemType`s. For the future we might want to parse the problem types individually and collect any unknown ones, before throwing an error. This could conflict with the current design as well.

File: autograder-api/src/main/java/de/firemage/autograder/api/AbstractLinter.java [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+import fluent.bundle.FluentBundle;
+
+import java.io.IOException;
+import java.nio.file.Path;
+import java.util.List;
+import java.util.Locale;
+import java.util.function.Consumer;
+
+public interface AbstractLinter {
+    List<? extends AbstractProblem> checkFile(Path file, JavaVersion version, CheckConfiguration checkConfiguration, Consumer<Translatable> statusConsumer) throws LinterException, IOException;
```
Comment (Luro02) [inline autograder-api/src/main/java/de/firemage/autograder/api/AbstractLinter.java:12]: Should we expose `JavaVersion`? Wouldn't this risk people forgetting about it and having a fixed version like `JavaVersion.JAVA_17` somewhere that would break when we update to Java 21?

And from what I can tell, one can always parse older java code with newer java versions, so it shouldn't be necessary to have this here. By the way, for the same reason, I want to remove the explicit java version set in every test as well. (Didn't do it yet, because it would touch hundreds of files)

File: autograder-api/src/main/java/de/firemage/autograder/api/AbstractLinter.java [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+    String translateMessage(Translatable translatable);
+    FluentBundle getFluentBundle();
```
Comment (Luro02) [inline autograder-api/src/main/java/de/firemage/autograder/api/AbstractLinter.java:14]: Same problem here, why do we expose the internals? Isn't `translateMessage` enough? If we want a method for explicitly specifying the language, we could have this instead?:
```java
default String translateMessage(Translatable translatable) {
    return this.transalteMessage(translatable, this.locale);
}

String translateMessage(Translatable translatable, Locale locale);
```

File: autograder-core/src/main/java/de/firemage/autograder/core/CodePosition.java [hunk 11]
```diff
@@ -55,7 +57,14 @@ public static CodePosition fromSourcePosition(
         );
     }
 
-    public String readString() {
+
+    @Override
+    public Path path() {
+        return this.file.toPath();
```
Comment (Luro02) [inline autograder-core/src/main/java/de/firemage/autograder/core/CodePosition.java:63]: Note that the path won't always be a real path. Some things are only in memory (I am not sure if I used somewhere other than the tests)

File: .github/workflows/build.yml (hunk 1 above)
Comment (Luro02) [reply .github/workflows/build.yml:65]: > Actually, the important reason is the thing that @dfuchss mentioned below: `autograder-full` may not contain `autograder-api`, or there will be clashes during dynamic loading. `autograder-cmd` on the other hand needs all dependencies.

From what I remember that is not a problem. The `URLClassLoader` will always ask the parent classloader first if it knows the class (in case it does, it will load from the parent), before it looks into the JAR. The `autograder-api` is embedded into the JAR, so it will always be known by the parent class loader

Comment (Luro02) [reply .github/workflows/build.yml:65]: That is the reason why I had to write the `DynamicClassLoader` which is a reimplementation of `URLClassLoader` with the above behaviour fixed.

File: autograder-api/src/main/java/de/firemage/autograder/api/AbstractLinter.java (hunk 10 above)
Comment (Feuermagier) [reply autograder-api/src/main/java/de/firemage/autograder/api/AbstractLinter.java:14]: You mean getFluentBundle(), right? I guess it can be removed; it was only needed at some point while implementing the PR

File: autograder-api/src/main/java/de/firemage/autograder/api/AbstractLinter.java (hunk 9 above)
Comment (Feuermagier) [reply autograder-api/src/main/java/de/firemage/autograder/api/AbstractLinter.java:12]: Not necessarily, since there may be checks (e.g. get(0) -> getFirst()) that are only applicable for specific Java versions

File: autograder-api/src/main/java/de/firemage/autograder/api/AbstractLinter.java (hunk 10 above)
Comment (Luro02) [reply autograder-api/src/main/java/de/firemage/autograder/api/AbstractLinter.java:14]: Yes

File: autograder-api/src/main/java/de/firemage/autograder/api/AbstractLinter.java (hunk 9 above)
Comment (Luro02) [reply autograder-api/src/main/java/de/firemage/autograder/api/AbstractLinter.java:12]: The autograder is currently not capable of having checks that only work with specific java versions. It also is very unlikely that we would downgrade from like Java 21 to 17 which would necessitate this.

Who knows what we do in the future, so we could keep it. Hopefully, nobody forgets changing that when we start using newer java versions

**Comment by dfuchss** [comment]
> Why is `ProblemType` in api? I guess the reason is for type-safety?
> 
> 
> 
> I wouldn't expose the enum to downstream, only via Strings for the enum constants.
> 
> This should prevent problems with potential future changes and the code using
> 
> `autograder-api` does not instantiate the `ProblemType` variants explicitly? It should
> 
> only parse the config, maybe not even that. If type-safety is necessary, then expose a wrapper class like this:
> 
> ```java
> 
> public record AutograderProblemType(String name) {}
> 
> 
> 
> // in `ProblemType` one could add a method to safely convert between them:
> 
> public enum ProblemType {
> 
>     ...;
> 
> 
> 
>     public AutograderProblemType toAutograderProblemType() {
> 
>         return new AutograderProblemType(this.toString());
> 
>     }
> 
> }
> 
> ```
> 
> 
> 
> The current implementation would crash when new releases introduce new checks because these are not loaded via the class loader. It would be exceedingly difficult to replace the `ProblemType` enum with a different implementation, like making it an interface and having different implementations for `autograder-core` and `autograder-extra`.
> 
> 
> 
> The current implementation doesn't provide a good error-message when there are unknown `ProblemType`s. For the future we might want to parse the problem types individually and collect any unknown ones, before throwing an error. This could conflict with the current design as well.

I'd agree if it changes often it should moved from the API and replaced by some interface for deserialization, we can provide the interface-class-relationship to Jackson . If it's kind of stable , it could stay here . That's something you should decide :)

**Comment by Feuermagier** [comment]
> Why is `ProblemType` in api? I guess the reason is for type-safety?
> 
> I wouldn't expose the enum to downstream, only via Strings for the enum constants. This should prevent problems with potential future changes and the code using `autograder-api` does not instantiate the `ProblemType` variants explicitly? It should only parse the config, maybe not even that. If type-safety is necessary, then expose a wrapper class like this:
> 
> ```java
> public record AutograderProblemType(String name) {}
> 
> // in `ProblemType` one could add a method to safely convert between them:
> public enum ProblemType {
>     ...;
> 
>     public AutograderProblemType toAutograderProblemType() {
>         return new AutograderProblemType(this.toString());
>     }
> }
> ```
> 
> The current implementation would crash when new releases introduce new checks because these are not loaded via the class loader. It would be exceedingly difficult to replace the `ProblemType` enum with a different implementation, like making it an interface and having different implementations for `autograder-core` and `autograder-extra`.
> 
> The current implementation doesn't provide a good error-message when there are unknown `ProblemType`s. For the future we might want to parse the problem types individually and collect any unknown ones, before throwing an error. This could conflict with the current design as well.

The current config format isn't going to be used with the new Artemis4J anyway, instead, the grading config and the autograder config will be merged (like [so](https://github.com/Feuermagier/artemis4j/blob/f63350672d211752c894e5336bb401e75b9b6c64/src/test/resources/config-newapi-test.json#L45)).

My intention with moving ProblemType to autograder-api was indeed type safety, though I see that this makes adding checks a lot harder, so I'm fine with using Strings only (@dfuchss also proposed this).

**Comment by dfuchss** [comment]
> > Why is `ProblemType` in api? I guess the reason is for type-safety?
> 
> > 
> 
> > I wouldn't expose the enum to downstream, only via Strings for the enum constants. This should prevent problems with potential future changes and the code using `autograder-api` does not instantiate the `ProblemType` variants explicitly? It should only parse the config, maybe not even that. If type-safety is necessary, then expose a wrapper class like this:
> 
> > 
> 
> > ```java
> 
> > public record AutograderProblemType(String name) {}
> 
> > 
> 
> > // in `ProblemType` one could add a method to safely convert between them:
> 
> > public enum ProblemType {
> 
> >     ...;
> 
> > 
> 
> >     public AutograderProblemType toAutograderProblemType() {
> 
> >         return new AutograderProblemType(this.toString());
> 
> >     }
> 
> > }
> 
> > ```
> 
> > 
> 
> > The current implementation would crash when new releases introduce new checks because these are not loaded via the class loader. It would be exceedingly difficult to replace the `ProblemType` enum with a different implementation, like making it an interface and having different implementations for `autograder-core` and `autograder-extra`.
> 
> > 
> 
> > The current implementation doesn't provide a good error-message when there are unknown `ProblemType`s. For the future we might want to parse the problem types individually and collect any unknown ones, before throwing an error. This could conflict with the current design as well.
> 
> 
> 
> The current config format isn't going to be used with the new Artemis4J anyway, instead, the grading config and the autograder config will be merged (like [so](https://github.com/Feuermagier/artemis4j/blob/f63350672d211752c894e5336bb401e75b9b6c64/src/test/resources/config-newapi-test.json#L45)).
> 
> 
> 
> My intention with moving ProblemType to autograder-api was indeed type safety, though I see that this makes adding checks a lot harder, so I'm fine with using Strings only (@dfuchss also proposed this). 

Just to make it clear: I really like the idea of having this enum in the API ; esp. because of TypeSafety in the Config Class. Nevertheless, you do know more about the frequency of changes to this enum :)

**Comment by Feuermagier** [comment]
It changes quite often (e.g. when we discover a frequent, easy to detect mistake during grading, and want to implement a new check for that). I'm going to use Strings instead of a wrapper class so that Artemis4J's config format is not going to depend on the presence of Jackson annotations within autograder-api

**Comment by dfuchss** [comment]
> > > Why is `ProblemType` in api? I guess the reason is for type-safety?
> > 
> > > 
> > 
> > > I wouldn't expose the enum to downstream, only via Strings for the enum constants. This should prevent problems with potential future changes and the code using `autograder-api` does not instantiate the `ProblemType` variants explicitly? It should only parse the config, maybe not even that. If type-safety is necessary, then expose a wrapper class like this:
> > 
> > > 
> > 
> > > ```java
> > 
> > > public record AutograderProblemType(String name) {}
> > 
> > > 
> > 
> > > // in `ProblemType` one could add a method to safely convert between them:
> > 
> > > public enum ProblemType {
> > 
> > >     ...;
> > 
> > > 
> > 
> > >     public AutograderProblemType toAutograderProblemType() {
> > 
> > >         return new AutograderProblemType(this.toString());
> > 
> > >     }
> > 
> > > }
> > 
> > > ```
> > 
> > > 
> > 
> > > The current implementation would crash when new releases introduce new checks because these are not loaded via the class loader. It would be exceedingly difficult to replace the `ProblemType` enum with a different implementation, like making it an interface and having different implementations for `autograder-core` and `autograder-extra`.
> > 
> > > 
> > 
> > > The current implementation doesn't provide a good error-message when there are unknown `ProblemType`s. For the future we might want to parse the problem types individually and collect any unknown ones, before throwing an error. This could conflict with the current design as well.
> > 
> > 
> > 
> > The current config format isn't going to be used with the new Artemis4J anyway, instead, the grading config and the autograder config will be merged (like [so](https://github.com/Feuermagier/artemis4j/blob/f63350672d211752c894e5336bb401e75b9b6c64/src/test/resources/config-newapi-test.json#L45)).
> > 
> > 
> > 
> > My intention with moving ProblemType to autograder-api was indeed type safety, though I see that this makes adding checks a lot harder, so I'm fine with using Strings only (@dfuchss also proposed this). 
> 
> Just to make it clear: I really like the idea of having this enum in the API ; esp. because of TypeSafety in the Config Class. Nevertheless, you do know more about the frequency of changes to this enum :)

I'd also would define a interface that is implemented by the enum instead of strings if it has to be moved. Thereby, the config can use the interface and the only thing to do is to provide Jackson with the information which enum class implements the interface

**Comment by Feuermagier** [comment]
How do you imagine the Jackson thing, given that the Enum is not available at compile time so it can't be used within annotations? Via a custom deserializer?

File: .github/workflows/build.yml (hunk 1 above)
Comment (Feuermagier) [reply .github/workflows/build.yml:65]: Seems like you're right, so I'm removing autograder-full

**Comment by Luro02** [comment]
Should we really have jackson in the interface? Couldn't we have something like this instead?:
```java
interface AbstractProblemType {
    AbstractProblemType fromString(String string);

    List<AbstractProblemType> fromStrings(List<String> strings);
}
```

**Comment by Feuermagier** [comment]
Looks like a good idea to me

**Comment by dfuchss** [comment]
> How do you imagine the Jackson thing, given that the Enum is not available at compile time so it can't be used within annotations? Via a custom deserializer?

Just define a simpleModule and define Mapping from Interface to ImplClass :) That should work. https://fasterxml.github.io/jackson-databind/javadoc/2.7/com/fasterxml/jackson/databind/module/SimpleModule.html#addAbstractTypeMapping(java.lang.Class,%20java.lang.Class)

File: .github/workflows/build.yml (hunk 1 above)
Comment (dfuchss) [reply .github/workflows/build.yml]: Are you sure about the loading order?
What happens if you modify the API e.g., add some classes & change some. Then the API partly consists of the bundled API and the dependend API. Probably, I'd suggest to make sure that the API is not delivered; just to be sure that it won't generate conflicts.

Comment (Feuermagier) [reply .github/workflows/build.yml]: Java guarantees the order:

> The ClassLoader class uses a delegation model to search for classes and resources. Each instance of ClassLoader has an associated parent class loader. When requested to find a class or resource, a ClassLoader instance will delegate the search for the class or resource to its parent class loader before attempting to find the class or resource itself.

That's from ClassLoader's Javadoc

Comment (Feuermagier) [reply .github/workflows/build.yml]: And your case would be broken anyway, since autograder-core and the downstream user would have the same name referring to different classes

**Review by Luro02 — APPROVED** [review]
